ref(control_silo): Consolidate integrations tasks and tasks' tests

Christinarlong opened the pull request:
Moves all integrations tasks and related tests to sentry/integrations/tasks or the specific integration provider/tasks (i.e github/tasks). Includes tasks shims for moved tasks.

ref(#73859)

Christinarlong committed:
ref(control_silo): Consolidate integrations tasks and tasks' tests

Moves all integrations tasks and related tests to sentry/integrations/tasks
or the specific integration provider/tasks (i.e github/tasks). Includes tasks shims
for moved tasks.

ref(#73859)

Christinarlong committed:
removed nonexsistant directory

Christinarlong committed:
updating celerybeat schedule and imports

markstory commented on src/sentry/integrations/tasks/sync_assignee_outbound_impl.py:
```diff
@@ -0,0 +1,47 @@
+
+
+@instrumented_task(
+    name="sentry.integrations.tasks.sync_assignee_outbound_impl",
```
Did this task name change? I don't see this name in any current tasks.

Christinarlong replied:
yes, in the previous/old module the task was named `sentry.tasks.integrations.sync_assignee_outbound`. In the new task/location it's updated to have the full module name. 

Will this lead to conflicts? It shouldn't (?), because all the new module paths are being added to the celery_imports

markstory replied:
There shouldn't be any conflicts as you've kept a task with the old name around to handle the deploy boundary. Since we're moving/renaming this task could we skip the `_impl` suffix? We have a different namespace which will give us unique names.

Christinarlong commented on src/sentry/integrations/api/bases/external_actor.py:
```diff
@@ -0,0 +1,165 @@
+}
+
+
+class ExternalActorSerializerBase(CamelSnakeModelSerializer):
```
agh I thought I got all of these. I think this is an artifact from making changes on one branch, moving branches, and then the changes get added to the current branches if a new file from the prev. branch gets updated.

Christinarlong committed:
merge Master into branch

Christinarlong committed:
remove unnecessary changes

markstory commented on src/sentry/integrations/slack/tasks/find_channel_id_for_rule.py:
```diff
@@ -0,0 +1,115 @@
+
+
+@instrumented_task(
+    name="sentry.integrations.slack.search_channel_id",
```
I'm seeing two tasks with this name. One in this file and the other in `src/sentry/tasks/integrations/slack/find_channel_id_for_rule`. We should have unique names for celery tasks.

markstory commented on src/sentry/integrations/slack/tasks/post_message.py:
```diff
@@ -0,0 +1,55 @@
+
+# TODO: add retry logic
+@instrumented_task(
+    name="sentry.integrations.slack.post_message_control",
```
This task and `sentry.integrations.slack.tasks.post_message` both have duplicate tasks registered. We could remove the task defined in `src/sentry/tasks/integrations/slack/post_message` and only do an import shim to avoid the duplicate task names.

Christinarlong replied:
what does duplicate task registered mean? Is that like two separate tasks are named the same ? 

Also for the import shim, would that look something like the class/symbol import shims ? I'm kinda confused rn, don't the task import shims need the `@instrumented_task` decorator to be valid?

markstory replied:
> what does duplicate task registered mean? Is that like two separate tasks are named the same ?

Yes, two tasks have the same `name` parameter provided to `instrumented_task`. Celery will get confused when trying to execute these tasks and we should avoid that.

markstory commented on src/sentry/integrations/tasks/kick_off_status_syncs_impl.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@instrumented_task(
+    name="sentry.integrations.tasks.kick_off_status_syncs_impl",
```
We could rename this task and remove the `_impl`. We'd need a shim task with the old task name for deploy compatibility.

Christinarlong replied:
`src/sentry/tasks/integrations/kick_off_status_syncs_impl.py` has a shim task with `"sentry.tasks.integrations.kick_off_status_syncs"` would that work for compatibility?

markstory replied:
Yeah that would work, I mostly want to avoid having `_impl` in the task name as that will require another round of renames to fix.

markstory commented on src/sentry/integrations/tasks/link_all_repos.py:
```diff
@@ -0,0 +1,88 @@
+
+
+@instrumented_task(
+    name="sentry.integrations.tasks.link_all_repos",
```
Should these tasks be within the `sentry.integrations.github` ? We have some tasks within their integration (some slack tasks) and other tasks are in `sentry.integrations`. Perhaps we can settle on having integration specific tasks within that integration?

Christinarlong replied:
sounds good to me 👍

markstory commented on src/sentry/integrations/tasks/migrate_issues.py:
```diff
@@ -0,0 +1,110 @@
+    max_retries=5,
+)
+@retry(exclude=(Integration.DoesNotExist))
+def migrate_issues(integration_id: int, organization_id: int) -> None:
```
We could put this task in the jira integration?

markstory commented on src/sentry/integrations/tasks/migrate_opsgenie_plugins.py:
```diff
@@ -0,0 +1,124 @@
+    max_retries=5,
+)
+@retry(exclude=(Integration.DoesNotExist, OrganizationIntegration.DoesNotExist))
+def migrate_opsgenie_plugin(integration_id: int, organization_id: int) -> None:
```
This module could be moved to `sentry.integrations.opsgenie.tasks` 🤷

markstory commented on src/sentry/integrations/tasks/sync_metadata.py:
```diff
@@ -0,0 +1,25 @@
+    silo_mode=SiloMode.CONTROL,
+)
+@retry(on=(IntegrationError,), exclude=(Integration.DoesNotExist,))
+def sync_metadata(integration_id: int) -> None:
```
Could move this task into the jira integration.

markstory commented on src/sentry/integrations/tasks/vsts/subscription_check.py:
```diff
@@ -0,0 +1,90 @@
+    silo_mode=SiloMode.CONTROL,
+)
+@retry(exclude=(ApiError, ApiUnauthorized, Integration.DoesNotExist))
+def vsts_subscription_check(integration_id: int, organization_id: int) -> None:
```
Same here with putting these tasks in `sentry.integrations.vsts.tasks`. We could skip renaming the task for now and do that in a followup change.

Christinarlong committed:
fix duplicate names and clearer seperation

Christinarlong commented on src/sentry/integrations/tasks/vsts/__init__.py:
```diff
@@ -0,0 +1,9 @@
+from sentry.integrations.tasks.vsts.kickoff_subscription_check import (
+    kickoff_vsts_subscription_check,
+)
+from sentry.integrations.tasks.vsts.subscription_check import vsts_subscription_check
```
Agreege, though there's an issue when having the vsts tasks in the vsts namespace (i.e sentry/integrations/vsts/tasks) where there's a cyclical dependency for the plugins module (1s lemme get a screenshot to remember what exactly was the problem)

Christinarlong replied:
Screenshot below. But it seems like the path is 
- setup -> plugins -> start loading other modules .... -> integrations/tasks/init.py -> integrations/vsts/init.py -> rules -> plugins
<img width="754" alt="image" src="https://github.com/user-attachments/assets/65af23f0-a29c-4ed7-96ab-51cd3ac9274a">

Christinarlong committed:
merge Master into branch

Christinarlong committed:
fix jira module names

Christinarlong committed:
Merge branch 'Master' into christinarlong/consolidate-integrations-tasks

Christinarlong committed:
moved all tasks to be under integrations/tasks

Christinarlong committed:
fix import paths in tests

Christinarlong committed:
no one saw that

markstory committed:
Fix up naming consistency

Christinarlong committed:
fix mypy errors

Christinarlong committed:
fix mypy errors

Christinarlong committed:
fix mypy errors

Christinarlong commented on src/sentry/integrations/tasks/sync_status_inbound.py:
```diff
@@ -0,0 +1,261 @@
+    if not affected_groups:
+        return
+
+    installation = integration.get_installation(organization_id=organization_id)
```
Would it be bad, to leave off the type hint for the return type of `get_installation()` ? Trying to import `IssueSyncMixin` in the `get_installation()` file (to typehint, IntegrationInstallation | IssueSyncMixin) causes an explosion of circular import errors. Currently alot of the mypy errors are solved by hasattr() which, doesn't seem as clean as saying,  installation is X | Y type.

Christinarlong commented on src/sentry/integrations/tasks/sync_status_inbound.py:
```diff
@@ -0,0 +1,261 @@
+            activity_type,
+            activity_data,
+        ) = get_resolutions_and_activity_data_for_groups(
+            list(affected_groups),
+            str(config.get("resolution_strategy")),
```
coerced for mypy

asottile-sentry replied:
`str(x.get(...))` is going to end up with `"None"` which is almost certainly wrong -- probably the correct response to this error should be `config['resolution_strategy']` -- or the caller should accept `None`?

Christinarlong replied:
gave caller `None`typehint for this param 👍

markstory committed:
Fix more mypy issues.

markstory committed:
Remove `tasks` from task names.

Having the order of the task name not match the module path was irking
me. While we'd like to change directory structure that work is blocked
by circular imports

markstory committed:
Fix incorrect logic caused by address types

Types were wrong, and logic was aligned with types when it was the types
that were wrong.

Christinarlong committed:
nits for tests

Christinarlong committed:
prayge last mypy fix

getsantry[bot] committed:
:knife: regenerate mypy module blocklist

Christinarlong committed:
update installed_apps

Christinarlong committed:
make get_resolutions accept None type

asottile-sentry commented on src/sentry/integrations/tasks/slack/send_notifications_on_activity.py:
```diff
@@ -10,15 +10,15 @@
 
 @instrumented_task(
-    name="sentry.integrations.slack.tasks.send_activity_notifications_to_slack_threads",
+    name="sentry.integrations.slack.send_activity_notifications_to_slack_threads",
```
I'm pretty sure changing the `name` of a task is a breaking change -- old code won't be able to queue up for the new code and new code won't be able to queue up for the old code (in the case of a rollback)

markstory replied:
Yes, changing celery task names will cause failures. I'll restore this and the other changed task names, as we don't want to drop any tasks during deploys. A rollback would cause failures with the changes as they stand right now, as we're introducing new task names that wouldn't be present in a previous release 🤔

asottile-sentry replied:
probably need a multi-stage rollout then (1) introduce new tasks (2) cut over to new tasks (3) remove old tasks

markstory replied:
Yeah. We'll split this up into multiple stages so that we can bridge rollbacks and forwards deploys.

markstory committed:
Revert "Remove `tasks` from task names."

This reverts commit c648f9b1959f22aa23ef85133ec02a2f8c8bf4cd.

Christinarlong commented:
Closing as PR has been split :^)
